### PR TITLE
Apply rich text bottom sheet [PART 1]

### PR DESCRIPTION
**Pull request by ManhNTX**
SpecialStyle:
https://user-images.githubusercontent.com/107173849/192460055-2169d76d-c889-42e8-b343-8569c1bafa3b.mov
QuickStyle:

https://user-images.githubusercontent.com/107173849/192658569-07591db6-9513-416c-90a2-7e53c8a078dd.mov

 ParagraphType, OrderListType ,In-OutDentType:

https://user-images.githubusercontent.com/107173849/192918872-f137c637-ec98-4ee2-9dc6-6cd924ffb702.mov

Picker TextColor and BackgroundColor:

https://user-images.githubusercontent.com/107173849/192934177-1fc79240-ed5f-446f-9a07-5d6eee226bf4.mov

### CODE REVIEW

File: pubspec.yaml [hunk 1]
```diff
@@ -12,7 +13,11 @@ dependencies:
     sdk: flutter
   flutter_keyboard_visibility: 5.2.0
   flutter_svg: 1.1.0
-
+  get: 4.6.5
+  enough_html_editor:
+    git:
+      url: https://github.com/ManhNTX/enough_html_editor.git
+      ref: improvements_for_use_tmail
```
Comment (hoangdat) [inline pubspec.yaml:20]: please create PR to linagora repository

**Comment by dab246** [comment]
Should close popup `format` after select one style format


Screen Shot 2022-09-27 at 14 44 40

**Comment by ManhNTX** [comment]
> Should close popup `format` after select one style format
> 
> Screen Shot 2022-09-27 at 14 44 40

User can select muti style format

**Comment by dab246** [comment]
> > Should close popup `format` after select one style format
> > Screen Shot 2022-09-27 at 14 44 40
> 
> User can select muti style format

@hoangdat  Should we do the same? IMO, it causes the user to take multiple actions.

File: lib/richtext_controller.dart [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+class RichTextController {
+
+  HtmlEditorApi? htmlEditorApi;
+
+  final listSpecialTextStyleApply = RxSet<SpecialStyleType>();
+  final headerStyleTypeApply = Rx<HeaderStyleType>(HeaderStyleType.normal);
+
+  final StreamController<bool> richTextStreamController = StreamController<bool>();
+
+  final isSpecialStyleSelected = [false, false, false, false];
```
Comment (dab246) [inline lib/richtext_controller.dart:119]: IMO, Using the variable list `bool` has no obvious meaning. You should create an object that wraps the `SpecialStyleType` object and has a `State` variable inside.

File: lib/views/widgets/option_bottom_sheet.dart [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+                    fit: BoxFit.fill,
+                    width: 24,
+                    height: 24,
+                    package: packageName,
+                  ),
+                ),
+              ],
+            ),
+          ),
+          Container(color: const Color(0xFFE4E4E4), height: 1),
```
Comment (dab246) [inline lib/views/widgets/option_bottom_sheet.dart:64]: Shoud add `color` in `CommonColor ` class for better management

File: example/pubspec.yaml [hunk 4]
```diff
@@ -36,6 +36,10 @@ dependencies:
   cupertino_icons: ^1.0.2
   rich_text_composer:
     path: ../
+  enough_html_editor:
+    git:
+      url: https://github.com/ManhNTX/enough_html_editor.git
```
Comment (dab246) [inline example/pubspec.yaml:41]: please use `linagora repository`

File: lib/views/widgets/rich_text_keyboard_toolbar.dart [hunk 5]
```diff
@@ -53,7 +60,21 @@ class RichTextKeyboardToolBar extends StatelessWidget {
...
+            await htmlEditorApi?.unfocus().then((_) {
+              showModalBottomSheet(
+                context: context,
+                shape: RoundedRectangleBorder(
+                  borderRadius: BorderRadius.circular(16),
+                ),
+                backgroundColor: Colors.white,
+                builder: (context) => RichTextOptionBottomSheet(
+                  richTextController: richTextController,
+                  title: 'Format',
```
Comment (dab246) [inline lib/views/widgets/rich_text_keyboard_toolbar.dart:73]: Please customize it to be able to support multi-language

File: lib/views/widgets/rich_text_option_bottom_sheet.dart [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+                      richTextController.headerStyleTypeApply.value = item;
+                    },
+                  ));
+        },
+        child: Row(
+          mainAxisSize: MainAxisSize.max,
+          mainAxisAlignment: MainAxisAlignment.center,
+          children: [
+            const Text(
+              'Quick styles',
```
Comment (dab246) [inline lib/views/widgets/rich_text_option_bottom_sheet.dart:142]: same

File: lib/views/widgets/rich_text_option_bottom_sheet.dart [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+          _buildIconButton(false, () {}, _imagePaths.icBulletOrder),
+          _buildVerticalDivider(),
+          _buildIconButton(false, () {}, _imagePaths.icNumberOrder),
+        ],
+      ),
+    );
+  }
+
+  Widget _buildVerticalDivider() => Container(
+        width: 1,
```
Comment (dab246) [inline lib/views/widgets/rich_text_option_bottom_sheet.dart:226]: remove indent

File: lib/views/widgets/rich_text_option_bottom_sheet.dart [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+          color: isSelected ? CommonColor.colorBackgroundSelect : Colors.white,
+          padding: const EdgeInsets.symmetric(vertical: 8, horizontal: 10),
+          height: double.infinity,
+          child: LimitedBox(
+            maxWidth: 28,
+            maxHeight: 28,
+            child: SvgPicture.asset(
+              asset,
+              color: isSelected ? CommonColor.colorIconSelect : null,
+              package: packageName,
```
Comment (dab246) [inline lib/views/widgets/rich_text_option_bottom_sheet.dart:248]: why need use it?

File: lib/views/widgets/list_header_style.dart [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+class ListHeaderStyle extends StatelessWidget {
+  const ListHeaderStyle({Key? key, required this.itemSelected})
+      : super(key: key);
+  final Function(HeaderStyleType) itemSelected;
+
+  @override
+  Widget build(BuildContext context) {
+    return OptionBottomSheet(
+      title: 'Quick styles',
```
Comment (dab246) [inline lib/views/widgets/list_header_style.dart:14]: Please customize it to be able to support multi-language

**Review by dab246 — COMMENTED** [review]
You should add `.lock` files to `.gitignore` to ignore it

File: lib/views/widgets/rich_text_option_bottom_sheet.dart (hunk 8 above)
Comment (ManhNTX) [reply lib/views/widgets/rich_text_option_bottom_sheet.dart:335]: Need it for load image form package without defining it on `pubspec.yaml` file of project implement

**Review by hoangdat — DISMISSED** [review]

**Review by hoangdat — APPROVED** [review]

File: lib/richtext_controller.dart [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+        listSpecialTextStyleApply.notifyListeners();
+      } else {
+        listSpecialTextStyleApply.value.remove(SpecialStyleType.strikeThrough);
+        isStrikeThroughAppended = false;
+        listSpecialTextStyleApply.notifyListeners();
+      }
+    };
+  }
+
+  selectTextStyleType(SpecialStyleType richTextStyleType) {
```
Comment (dab246) [inline lib/richtext_controller.dart:75]: If there is no function return, you should use the `void` keyword before the function name.

File: lib/richtext_controller.dart [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+    if (listSpecialTextStyleApply.value.contains(richTextStyleType)) {
+      listSpecialTextStyleApply.value.remove(richTextStyleType);
+    } else {
+      listSpecialTextStyleApply.value.add(richTextStyleType);
+    }
+    listSpecialTextStyleApply.notifyListeners();
+  }
+
+  selectTextColor(ui.Color color) {
```
Comment (dab246) [inline lib/richtext_controller.dart:84]: same

File: lib/richtext_controller.dart [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+    selectedTextColor.value = color;
+  }
+
+  selectBackgroundColor(ui.Color color) {
```
Comment (dab246) [inline lib/richtext_controller.dart:88]: same

File: lib/richtext_controller.dart [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+
+  Future<void> applyTextColor() async {
+    htmlEditorApi?.setColorTextForeground(selectedTextColor.value);
+  }
+
+  Future<void> applyBackgroundTextColor() async {
+    htmlEditorApi?.setColorTextBackground(selectedTextBackgroundColor.value);
+  }
+
+  selectOrderListType(OrderListType orderListType) {
```
Comment (dab246) [inline lib/richtext_controller.dart:183]: same

File: lib/richtext_controller.dart [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+  void showRichTextView() {
+    richTextStreamController.sink.add(true);
+  }
+
+  void hideRichTextView() {
+    richTextStreamController.sink.add(false);
+  }
+
+  Future<void> editorOnFocus() async {
+    await applyParagraphType();
```
Comment (dab246) [inline lib/richtext_controller.dart:215]: These functions can run in parallel. Why don't you use `Future.wait()`

File: lib/richtext_controller.dart [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+    selectedTextColor.value = color;
+  }
+
+  selectBackgroundColor(ui.Color color) {
+    selectedTextBackgroundColor.value = color;
+  }
+
+  bool isTextStyleTypeSelected(SpecialStyleType richTextStyleType) {
+    return listSpecialTextStyleApply.value.contains(richTextStyleType);
+
```
Comment (dab246) [inline lib/richtext_controller.dart:94]: remove line

File: lib/views/commons/colors.dart [hunk 16]
```diff
@@ -0,0 +1,11 @@
+import 'package:flutter/material.dart';
+
+class CommonColor {
+ static const colorBorderGray = Color(0xFFE4E4E4);
+ static const colorIconSelect = Color(0xFF99A2AD);
+ static const colorBackgroundSelect = Color(0xFFF2F3F5);
+ static const colorStyleBlockQuote = Color(0xFFEEEEEE);
+ static const colorBorderStyleCode = Color(0xFFCCCCCC);
+ static const colorBackgroundStyleCode = Color(0xFFF5F5F5);
+
```
Comment (dab246) [inline lib/views/commons/colors.dart:10]: remove line

File: lib/views/widgets/list_color.dart [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+      padding: const EdgeInsets.all(0),
+      child: Column(
+        children: [
+          const Spacer(),
+          SizedBox(
+            height: _kKeyboardHeight,
+            child: Wrap(
+              direction: Axis.vertical,
+              children: <Widget>[
+                for (final color in listColor)
```
Comment (dab246) [inline lib/views/widgets/list_color.dart:35]: should use `list.map`. Ex:
```dart
Wrap(
    direction: Axis.vertical,
    children: listColor.map((color) {
       return GestureDetector(
            onTap: () => onSelected.call(color),
            child: Container(
              color: color,
              width: itemWidth,
              height: itemHeight,
            ),
                  );
    }).toList()
)
```

**Review by dab246 — COMMENTED** [review]
Please test `ColorPicker` on real device

**Comment by ManhNTX** [comment]
> Please test `ColorPicker` on real device

Tested. Its OK

File: lib/richtext_controller.dart [hunk 18]
```diff
@@ -0,0 +1,243 @@
...
+  bool isStrikeThroughAppended = false;
+
+  Stream<bool> get richTextStream => richTextStreamController.stream;
+
+  void listenHtmlEditorApi() {
+    htmlEditorApi?.onFormatSettingsChanged = (formatSettings) {
+      if (formatSettings.isBold) {
+        isBoldStyleAppended = true;
+        listSpecialTextStyleApply.value.add(SpecialStyleType.bold);
+        listSpecialTextStyleApply.notifyListeners();
```
Comment (dab246) [inline lib/richtext_controller.dart:38]: IMO, you don't need to call it. The `ValueNotifier ` variable will dynamically call `notifyListeners ` when you change its value.


Screen Shot 2022-09-30 at 13 04 40

File: lib/richtext_controller.dart [hunk 19]
```diff
@@ -0,0 +1,243 @@
...
+      } else {
+        listSpecialTextStyleApply.value.remove(SpecialStyleType.bold);
+        isBoldStyleAppended = false;
+        listSpecialTextStyleApply.notifyListeners();
```
Comment (dab246) [inline lib/richtext_controller.dart:42]: same

File: lib/richtext_controller.dart [hunk 20]
```diff
@@ -0,0 +1,243 @@
...
+      }
+
+      if (formatSettings.isItalic) {
+        isItalicStyleAppended = true;
+        listSpecialTextStyleApply.value.add(SpecialStyleType.italic);
+        listSpecialTextStyleApply.notifyListeners();
```
Comment (dab246) [inline lib/richtext_controller.dart:48]: same

File: lib/richtext_controller.dart [hunk 21]
```diff
@@ -0,0 +1,243 @@
...
+      } else {
+        listSpecialTextStyleApply.value.remove(SpecialStyleType.italic);
+        isItalicStyleAppended = false;
+        listSpecialTextStyleApply.notifyListeners();
```
Comment (dab246) [inline lib/richtext_controller.dart:52]: same

File: lib/richtext_controller.dart [hunk 22]
```diff
@@ -0,0 +1,243 @@
...
+      }
+
+      if (formatSettings.isUnderline) {
+        isUnderlineAppended = true;
+        listSpecialTextStyleApply.value.add(SpecialStyleType.underline);
+        listSpecialTextStyleApply.notifyListeners();
```
Comment (dab246) [inline lib/richtext_controller.dart:58]: same

File: lib/richtext_controller.dart [hunk 23]
```diff
@@ -0,0 +1,243 @@
...
+      } else {
+        listSpecialTextStyleApply.value.remove(SpecialStyleType.underline);
+        isUnderlineAppended = false;
+        listSpecialTextStyleApply.notifyListeners();
```
Comment (dab246) [inline lib/richtext_controller.dart:62]: same

File: lib/richtext_controller.dart [hunk 24]
```diff
@@ -0,0 +1,243 @@
...
+      }
+
+      if (formatSettings.isStrikeThrough) {
+        isStrikeThroughAppended = true;
+        listSpecialTextStyleApply.value.add(SpecialStyleType.strikeThrough);
+        listSpecialTextStyleApply.notifyListeners();
+      } else {
+        listSpecialTextStyleApply.value.remove(SpecialStyleType.strikeThrough);
+        isStrikeThroughAppended = false;
+        listSpecialTextStyleApply.notifyListeners();
```
Comment (dab246) [inline lib/richtext_controller.dart:72]: same

File: lib/richtext_controller.dart [hunk 25]
```diff
@@ -0,0 +1,243 @@
...
+      } else {
+        listSpecialTextStyleApply.value.remove(SpecialStyleType.underline);
+        isUnderlineAppended = false;
+        listSpecialTextStyleApply.notifyListeners();
+      }
+
+      if (formatSettings.isStrikeThrough) {
+        isStrikeThroughAppended = true;
+        listSpecialTextStyleApply.value.add(SpecialStyleType.strikeThrough);
+        listSpecialTextStyleApply.notifyListeners();
```
Comment (dab246) [inline lib/richtext_controller.dart:68]: same

File: lib/richtext_controller.dart [hunk 26]
```diff
@@ -0,0 +1,243 @@
...
+    };
+  }
+
+  void selectTextStyleType(SpecialStyleType richTextStyleType) {
+    if (listSpecialTextStyleApply.value.contains(richTextStyleType)) {
+      listSpecialTextStyleApply.value.remove(richTextStyleType);
+    } else {
+      listSpecialTextStyleApply.value.add(richTextStyleType);
+    }
+    listSpecialTextStyleApply.notifyListeners();
```
Comment (dab246) [inline lib/richtext_controller.dart:83]: same

**Review by dab246 — COMMENTED** [review]
@ManhNTX  Please change min flutter sdk is `2.17.0` to avoid conflicts when integrating into `TMail`

File: lib/richtext_controller.dart (hunk 18 above)
Comment (ManhNTX) [reply lib/richtext_controller.dart:38]: @dab246 IMO, The `ValueNotifier` variable will dynamically call `notifyListeners` when you change its value.  But when I added a new object to the list, the list will not change so `notifyListeners` will not call.

**Review by dab246 — APPROVED** [review]
/lgtm

**Review by hoangdat — APPROVED** [review]

**Pull request merged**